provider/gateway/rest/router.go

```diff
@@ -178,6 +183,86 @@ func newJwtServerRouter(addr sdk.Address, privateKey interface{}, jwtExpiresAfte
 	return router
 }
 
+func newResourceServerRouter(log log.Logger, providerAddr sdk.Address, publicKey *ecdsa.PublicKey, lokiGwAddr string) *mux.Router {
+	router := mux.NewRouter()
+
+	// add a middleware to verify the JWT provided in Authorization header
+	router.Use(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// verify the provided JWT
+			token, err := jwt.ParseWithClaims(r.Header.Get("Authorization"), &ClientCustomClaims{}, func(token *jwt.Token) (interface{}, error) {
+				// return the public key to be used for JWT verification
+				return publicKey, nil
+			})
+			if err != nil {
+				log.Error(err.Error())
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+			// delete the Authorization header as it is no more needed
+			r.Header.Del("Authorization")
+
+			// store the owner & provider address in request context to be used in later handlers
+			ownerAddress, err := sdk.AccAddressFromBech32(token.Claims.(*ClientCustomClaims).Subject)
+			if err != nil {
+				log.Error(err.Error())
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+			gcontext.Set(r, ownerContextKey, ownerAddress)
+			gcontext.Set(r, providerContextKey, providerAddr)
+
+			next.ServeHTTP(w, r)
+		})
+	})
+
+	lrouter := router.PathPrefix(leasePathPrefix).Subrouter()
+	lrouter.Use(requireLeaseID())
+
+	lokiServiceRouter := lrouter.PathPrefix("/loki-service").Subrouter()
+	lokiServiceRouter.NewRoute().Handler(lokiServiceHandler(log, lokiGwAddr))
+
+	return router
+}
+
+// lokiServiceHandler forwards all requests to the loki instance running in provider's cluster.
+// Example:
+// 		Incoming Request: http://localhost:8445/lease/1/1/1/loki-service/loki/api/v1/query?query={app=".+"}
+// 		Outgoing Request: http://{lokiGwAddr}/loki/api/v1/query?query={app=".+"}
+func lokiServiceHandler(log log.Logger, lokiGwAddr string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// set the X-Scope-OrgID header for fetching logs for the right tenant
+		r.Header.Set("X-Scope-OrgID", builder.LidNS(requestLeaseID(r)))
+
+		// build target url for the reverse proxy
+		scheme := "http" // for http & https
+		if strings.HasPrefix(r.URL.Scheme, "ws") {
+			scheme = "ws" // for ws & wss
+		}
+		lokiURL, err := url.Parse(fmt.Sprintf("%s://%s", scheme, lokiGwAddr))
```

Comment:
We should encapsulate all requests to loki into a client interface that can be defined in `provider/cluster/kube/loki`. That way it is easier to stub out in test, easier to establish a consist pattern of interaction with the 3rd party service, and easier to refactor in the future if we need to.

Reply:
If it is required that only a specific subset of loki APIs be exposed to the tenants, then a NGINX ingress can be put in-between the loki instance and the resource server with API whitelisting. LOKI_GATEWAY_ADDR will then point to the address of the ingress.

I think, this would be a better approach, than to define the loki interface in code as we can't control the loki version that provider's would use. And, each version can have different APIs exposed. So, better to let the provider's make this decision on their own.

Reply:
I'm ok with using nginx to make the API simpler, but we will have to document it well so that people don't run with it wide open by default.

Something we can address later.